File: content/blog/2003/atom/normalise_html.py
import sys
import urlparse

import libxml2

# turn off errors
def noerr(ctx, str): pass
libxml2.registerErrorHandler(noerr, None)

def parse(data, encoding='ISO-8859-1'):
    '''Parses an HTML fragment, and returns a <div> node containing the
    content in a DOM tree.  This makes it easy to insert into an XML
    document, and also gets rid of problems with unclosed elements.'''

    html_template = '<html><head><title></title></head><body>%s</body></html>'

    doc = libxml2.htmlParseDoc(html_template % data, encoding)
    # body is second element of root node
    body = doc.getRootElement().children.next

    div = libxml2.newNode('div')
    div.setNs(div.newNs('http://www.w3.org/1999/xhtml', None))

    # move nodes over to div
    child = body.children
    while child:
        nextchild = child.next
        child.unlinkNode()
        div.addChild(child)
        child = nextchild
    #doc.free()

    return div

def resolve_relative(node, base_uri):
    '''Makes sure that the URIs in a document are all absolute'''
    for attr in ('src', 'href', 'cite'):
        if node.hasProp(attr):
            node.setProp(attr, urlparse.urljoin(base_uri, node.prop(attr)))

    child = node.children
    while child:
        if child.type == 'element':
            resolve_relative(child, base_uri)
        child = child.next

def sanitise(node, remove_styles=False):
    '''Removes script elements and on* attributes.  Optionally removes
    style related elements and attributes too.

    This is not a fool proof way to sanitise content -- for that, you
    would want a complete list of allowed elements and attributes.'''
    
    # get rid of js handler attributes
    attr = node.properties
    while attr:
        print attr.name
        nextattr = attr.next
        if attr.name.lower().startswith('on') or \
               remove_styles and attr.name.lower() == 'style':
            attr.unlinkNode()
            attr.freeNode()
        attr = nextattr

    child = node.children
    while child:
        nextchild = child.next
        if child.type == 'element':
            if child.name == 'script' or \
                   remove_styles and (child.name == 'style' or
                                      (child.name == 'link' and
                                       child.hasProp('rel') and
                                       child.prop('rel').lower() == 'stylesheet')):
                child.unlinkNode()
                child.freeNode()
            else:
                sanitise(child, remove_styles=remove_styles)
        child = nextchild

doc = libxml2.newDoc('1,0')
div = parse(sys.stdin.read())
resolve_relative(div, 'http://www.gnome.org/~james/')
sanitise(div)
doc.addChild(div)

doc.formatDump(sys.stdout, True)

File: content/blog/2003/atom/normalise_html.c
#include <stdlib.h>
#include <libxml/parser.h>
#include <libxml/tree.h>
#include <libxml/HTMLparser.h>
#include <libxml/uri.h>

static void
ignore_errors(void *ctx, const char *msg, ...)
{
    /* nothing */
}

xmlNode *
parse_html_fragment(const char *buffer, size_t len, xmlCharEncoding encoding)
{
    /* wrapper for HTML fragment */
    static const char header[] = "<html><head><title></title></head><body>";
    static const char footer[] = "</body></html>";

    htmlParserCtxt *ctx;
    xmlDoc *htmldoc;
    xmlNode *body, *div, *child;
    xmlNs *xhtmlns;

    ctx = htmlCreatePushParserCtxt(NULL, NULL, NULL, 0, "foo", encoding);
    if (!ctx) {
	/* XXXX - could not create parser */
	return NULL;
    }
    xmlSetGenericErrorFunc(ctx, ignore_errors);
    /* do best effort to parse document */
    if (htmlParseChunk(ctx, header, sizeof(header)-1, 0) == 0) {
	if (htmlParseChunk(ctx, buffer, len, 0) == 0) {
	    htmlParseChunk(ctx, footer, sizeof(footer)-1, 1);
	}
    }
    /* get document from buffer */
    htmldoc = ctx->myDoc;
    htmlFreeParserCtxt(ctx);

    /* body is second child of root element */
    body = xmlDocGetRootElement(htmldoc)->children->next;

    div = xmlNewNode(NULL, "div");
    xhtmlns = xmlNewNs(div, "http://www.w3.org/1999/xhtml", NULL);
    xmlSetNs(div, xhtmlns);

    child = body->children;
    while (child) {
	xmlNode *nextchild = child->next;

	xmlUnlinkNode(child);
	xmlAddChild(div, child);
	child = nextchild;
    }
    xmlFreeDoc(htmldoc);

    return div;
}

void
resolve_relative(xmlNode *node, const char *base_uri)
{
    xmlNode *child;

    /* HTML spec mentioned 3 attributes on elements that we might want
     * to update */
    if (xmlHasProp(node, "src")) {
	xmlChar *uri = xmlBuildURI(xmlGetProp(node, "src"), base_uri);
	xmlSetProp(node, "src", uri);
	xmlFree(uri);
    }
    if (xmlHasProp(node, "href")) {
	xmlChar *uri = xmlBuildURI(xmlGetProp(node, "href"), base_uri);
	xmlSetProp(node, "href", uri);
	xmlFree(uri);
    }
    if (xmlHasProp(node, "cite")) {
	xmlChar *uri = xmlBuildURI(xmlGetProp(node, "cite"), base_uri);
	xmlSetProp(node, "cite", uri);
	xmlFree(uri);
    }

    for (child = node->children; child != NULL; child = child->next) {
	if (child->type == XML_ELEMENT_NODE)
	    resolve_relative(child, base_uri);
    }
}

xmlBuffer *
read_file(FILE *fp)
{
    xmlBuffer *ret;
    xmlChar buf[4096];
    size_t count;

    ret = xmlBufferCreate();
    if (!ret) return NULL;
    while ((count = fread(buf, sizeof(xmlChar), sizeof(buf), fp)) != 0) {
	xmlBufferAdd(ret, buf, count);
    }
    return ret;
}

int
main(int argc, char **argv)
{
    xmlBuffer *buf;
    xmlNode *div;
    xmlDoc *doc;

    /* parse stdin as an HTML fragment */
    buf = read_file(stdin);
    div = parse_html_fragment(xmlBufferContent(buf), xmlBufferLength(buf),
			      XML_CHAR_ENCODING_UTF8);

    /* optionally resolve relative URIs */
    if (argc >= 2)
	resolve_relative(div, argv[1]);

    /* embed parsed content into an XML document */
    doc = xmlNewDoc("1.0");
    xmlDocSetRootElement(doc, div);

    /* output */
    xmlDocFormatDump(stdout, doc, 1);

    xmlFreeDoc(doc);
    return 0;
}
